### PR TITLE
[min-cut partitioner] Disable a heuristic if graph has recomputable ops

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103601
* __->__ #103635

Removing this heuristic leads to major memory compression and speedup bump for activation-checkpointed models. Here is the data
image

### CODE REVIEW

File: torch/_functorch/partitioners.py [hunk 1]
```diff
@@ -783,8 +783,12 @@ def ban_recomputation(node):
             # modification appears to have made this heuristic a lot less critical
             # for performance.
             # TODO: Investigate why this hack helps.
-            if compiler == "inductor" and node.dist_from_bw > config.max_dist_from_bw:
-                return True
+            # TODO: Investigate the interaction with compiler assisted
+            # activation checkpointing. Removing the heuristic improves both
+            # memory footprint and speedup.
+            if not graph_has_recomputable_ops:
```
Comment (Chillee) [inline torch/_functorch/partitioners.py:789]: Can we not just do the same thing where we apply it to each node individually and not the entire graph at once?

**Review by Chillee — APPROVED** [review]
Will approve for now, but I would like a resolution to the comment.

File: torch/_functorch/partitioners.py (hunk 1 above)
Comment (anijain2305) [reply torch/_functorch/partitioners.py:789]: I don't think we can apply per node here. This line is reached only for those nodes that do not have recompute tags (see L760 below).

https://github.com/pytorch/pytorch/blob/2adfd1315aaf45a356a8570f89054a6736aebb1f/torch/_functorch/partitioners.py#L760-L787

This PR is saying that for nodes sitting outside of activation checkpointing, lets disable the heuristic for them, and let min-cut partitioner find the cut. I looked at couple of models, and the saved tensor size reduction is quite large (like 30% - 50%).

Comment (Chillee) [reply torch/_functorch/partitioners.py:789]: Ok, I don't understand why this helps (but tbh, i also don't have a great understanding why the current heuristic helps). Would be nice to resolve those at some point.

Comment (anijain2305) [reply torch/_functorch/partitioners.py:789]: Yes, I tried to understand it as well. But wasn't able to see this on a small model. On large models, it gets difficult to reason why this works. 

For one model, it was pretty clear though. There was some `full` operator which was producing a large tensor, and we were stashing some intermediate tensor far away from that `full` op. With heuristic disabled, it went all the way to the full and saved memory footprint.

**Pull request closed without merging**